File: clojure/eu002.clj
(defn next-fib 
  "Computes next number in the fibonacci sequence,
  returns next and previous as list.
  e.g. (next-fib '(2 1)) -> '(3 2)"
  [xs] 
  (list (+ (first xs) (second xs)) (first xs)))

(defn even-fib-sum
  "Computes sum of all even-numbered members of the 
  fibonacci sequence below max-n"
  [max-n]
  (loop [fibs '(2 1)
         sum 0]
    (if (> (first fibs) max-n)
      sum
      (recur (next-fib fibs) 
             (if (even? (first fibs))
               (+ sum (first fibs))
               sum)))))

(def run-test '(= (even-fib-sum 50) 44))
(def run-problem '(even-fib-sum 4e6))
;(even-fib-sum 4e6) ;-> 4613732

File: julia/eu002.jl
#!/usr/bin/env julia

# fibsum computes the sum of all even-values members of the
# Fibonacci sequence below n.
function fibsum(n::Int)
    f0 = 1
    f1 = 2
    fnext = 3
    r = 2
    while f1 < n
        fnext = f0+f1
        if fnext < n && fnext%2 == 0
            r += fnext
        end
        f0 = f1
        f1 = fnext
    end
    r
end

using Test
@test fibsum(90) == 44

# problem
println(fibsum(4000000))

